armsrc/lfops.c

```diff
@@ -1286,43 +1355,132 @@ void T55xxResetRead(void) {
 	LED_A_OFF();
 }
 
-// Write one card block in page 0, no lock
-void T55xxWriteBlockExt(uint32_t Data, uint32_t Block, uint32_t Pwd, uint8_t arg) {
-	LED_A_ON();
-	bool PwdMode = arg & 0x1;
-	uint8_t Page = (arg & 0x2)>>1;
-	bool testMode = arg & 0x4;
-	uint32_t i = 0;
+// Send one downlink command to the card 
+void T55xx_SendCMD (uint32_t Data, uint32_t Block, uint32_t Pwd, uint8_t arg) { //, bool read_cmd) {//, struct T55xx_Timing *Timing) {
+	
+	/*
+		arg bits
+		xxxxxxx1 0x01 PwdMode
+		xxxxxx1x 0x02 Page
+		xxxxx1xx 0x04 testMode
+		xxx11xxx 0x18 downlink mode
+		xx1xxxxx 0x20 reg_readmode
+		x1xxxxxx 0x40 called for a read, so no data packet
+		
+	*/
+	bool 	PwdMode 		= ((arg & 0x01) == 0x01);
+	uint8_t Page 			=  (arg & 0x02) >> 1;
+	bool 	testMode		= ((arg & 0x04) == 0x04);
+	uint8_t downlink_mode   =  (arg >> 3) & 0x03;;
+	bool	reg_readmode	= ((arg & 0x20) == 0x20);
+	bool	read_cmd		= ((arg & 0x40) == 0x40);
+	
+	int i = 0;             
+	bool BitStream[100];  // Max Downlink Command size ~75 bits, so 10 bytes (80 bits)
+	uint8_t BitStreamLen;
```

Comment:
While the bool makes the loops simpler, I think you used less arm memory before (pwpiwi, correct me if I'm wrong) 
I believe bools are stored as bytes.

I'm not sure where we stand on arm memory usage currently.

Reply:
I did think about this a bit.  Happy to take advice.   When we talk memory are we talking about ram or flash.  I wondered about the array size v code size v speed (i.e. the math and byte/bit index size and shift operations).  I assumed that RAM space would only be taken while the code was getting run, then released.

Let me know which way and I will adjust.

Reply:
`bool` is stored as `int`. The `BitStream[100]` therefore temporarily requires 400 Bytes RAM. With 64k RAM available and a significant amount reserved by RAMFUNCs, `BigBuf[]` and `ToSend[]`, we should indeed be carefull. The code should work with uint8_t as well. Speed should not be an issue here.

Reply:
int (4 bytes) was not expecting that.  
Based on that, the 400 bytes RAM array will now be 80 bytes. In my dev I have converted back to uint8_t BitStream[10];  (i.e. 80 bits for a max of 74 needed)
I also improved the index math a little, so a bit cleaner (using more bitwise then math) then the  original.  So the best of both worlds
e.g.
	if (downlink_mode ==  T55xx_DLMode_1of4) { // 1 of 4 need to send 2 bits at a time
		for ( i = 0; i < BitStreamLen; i+=2 ) {
			SendBits    = (BitStream[ i     /8] >> ( i    & 0xff)) << 1;   // Bit i
                        SendBits += (BitStream[(i+1)/8] >> ((i+1) & 0xff));        // Bit i+1; 
			T55xxWriteBit (SendBits,Timing);
		}
	}
	else {
		for (i = 0; i < BitStreamLen; i++) {
			SendBits = BitStream[i/8] >> (i & 0xff) ; 
			T55xxWriteBit (SendBits,Timing);
    	     }		
	}
}

Does that sound OK ?
Anything else need looking at (in the current code) ?

Reply:
quick note for reference.  I run a sizeof for bool, uint_8 and int
#db# bool 1, uint_8 1, int 4
So looks like bool is 1 byte

Reply:
Fork Updated with the bit_stream using uint_8 array 10 bytes (80 bits)
Some int variables changed to uint8_t (no need to be bigger)
Defined 2 macros for better code layout/readability for the bit_stream read/write
moved some code to group new code together.

Not related to my code (do you want me to update to %llu ?
fpga_compress.c:102:46: warning: format '%lu' expects argument of type 'long unsigned int', but argument 3 has type 'long long unsigned int' [-Wformat=]
     fprintf(stderr, "Input file too big (> %lu bytes). This is probably not a hardnested bitflip state table.\n", HARDNESTED_TABLE_SIZE);
                                            ~~^
                                            %llu
and
fido/cbortools.c:213:3: warning: format '%d' expects argument of type 'int', but argument 3 has type 'long long int' [-Wformat=]
   "CBOR parsing failure at offset %" PRId32 " : %s\n",
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
fido/cbortools.c:215:3:
   cb.ptr - data, cbor_error_string(err));
   ~~~~~~~~~~~~~

Reply:
I stand corrected regarding the sizeof(bool). @mwalker33 is correct.

Reply:
Regarding the printf format strings: you are probably compiling 64bit code? On my machine the argument 3 (in fpga_compress.c) is 'long unsigned int'. Possible solution might be to add a `(size_t)` to the definition of `HARDNESTED_TABLE_SIZE` and to use `%zu` in the format string. But this should go into another Pull Request.

In cbortools.c I would propose to get rid of the `#if __WORDSIZE == 64` (it obviously isn't defined or isn't defined correctly) and use some standard constructs. There is a `%td` format string for pointer differences (like `cb.ptr - data`). But I am not sure if all compilers support that. Again, that would be another PR.

Without and with these proposed changes I don't get any warnings on Windows compiling 32bit code with gcc 6.3.0.

Reply:
OK, they are just warnings and don't seem to be an issue.  Just thought I would report and check in.
I will leave them as is.